File: examples/example_crm.cpp
// yaml
#include <yaml-cpp/yaml.h>

// fmt
#include <fmt/format.h>

// kintera
#include <kintera/constants.h>

#include <kintera/kinetics/evolve_implicit.hpp>
#include <kintera/kinetics/kinetics.hpp>
#include <kintera/kinetics/kinetics_formatter.hpp>
#include <kintera/thermo/relative_humidity.hpp>

// snap
#include <snap/input/command_line.hpp>
#include <snap/mesh/mesh_formatter.hpp>
#include <snap/mesh/meshblock.hpp>
#include <snap/output/output_formats.hpp>

using namespace snap;

int main(int argc, char **argv) {
  // read parameters
  auto cli = CommandLine::ParseArguments(argc, argv);
  if (!cli) return 0;

  // input file
  auto infile = std::string(cli->input_filename);
  auto device = torch::kCPU;
  if (torch::cuda::is_available()) {
    std::cout << "Running on CUDA" << std::endl;
    device = torch::kCUDA;
  }

  // experiment name is before "."
  auto exp_name = infile.substr(0, infile.find('.'));

  auto config = YAML::LoadFile(infile);
  auto Ps = config["problem"]["Ps"].as<double>(1.e5);
  auto Ts = config["problem"]["Ts"].as<double>(300.);
  auto Tmin = config["problem"]["Tmin"].as<double>(200.);
  auto grav = -config["forcing"]["const-gravity"]["grav1"].as<double>();

  // initialize the block
  auto block = MeshBlock(MeshBlockOptions::from_yaml(infile));
  std::cout << fmt::format("{}", block->options) << std::endl;
  block->to(device);

  // useful modules
  auto phydro = block->phydro;
  auto pcoord = phydro->pcoord;
  auto peos = phydro->peos;
  auto m = block->named_modules()["hydro.eos.thermo"];
  auto thermo_y = std::dynamic_pointer_cast<kintera::ThermoYImpl>(m);

  // dimensions and indices
  int nc3 = pcoord->x3v.size(0);
  int nc2 = pcoord->x2v.size(0);
  int nc1 = pcoord->x1v.size(0);
  int ny = thermo_y->options.species().size() - 1;
  int nvar = peos->nvar();

  // construct an adiabatic atmosphere
  kintera::ThermoX thermo_x(thermo_y->options);
  thermo_x->to(device);

  auto temp =
      Ts *
      torch::ones({nc3, nc2},
                  torch::TensorOptions().dtype(torch::kDouble).device(device));

  auto pres =
      Ps *
      torch::ones({nc3, nc2},
                  torch::TensorOptions().dtype(torch::kDouble).device(device));

  auto xfrac =
      torch::zeros({nc3, nc2, 1 + ny},
                   torch::TensorOptions().dtype(torch::kDouble).device(device));

  auto w = torch::zeros(
      {nvar, nc3, nc2, nc1},
      torch::TensorOptions().dtype(torch::kFloat64).device(device));

  // read in compositions
  for (int i = 1; i <= ny; ++i) {
    auto name = thermo_y->options.species()[i];
    auto xmixr = config["problem"]["x" + name].as<double>(0.);
    xfrac.select(2, i) = xmixr;
  }

  // dry air mole fraction
  xfrac.select(2, 0) = 1. - xfrac.narrow(-1, 1, ny).sum(-1);

  // adiabatic extrapolate half a grid to cell center
  int is = pcoord->is();
  int ie = pcoord->ie();
  auto dz = pcoord->dx1f[is].item<double>();
  std::cout << fmt::format("{}\n", Func1Registrar::list_names()) << std::endl;
  thermo_x->extrapolate_ad(temp, pres, xfrac, grav, dz / 2.);

  int i = is;
  int nvapor = thermo_x->options.vapor_ids().size();
  int ncloud = thermo_x->options.cloud_ids().size();
  for (; i <= ie; ++i) {
    auto conc = thermo_x->compute("TPX->V", {temp, pres, xfrac});

    w[IPR].select(2, i) = pres;
    w[IDN].select(2, i) = thermo_x->compute("V->D", {conc});

    auto result = thermo_x->compute("X->Y", {xfrac});
    w.narrow(0, ICY, ny).select(3, i) = thermo_x->compute("X->Y", {xfrac});

    if ((temp < Tmin).any().item<double>()) break;
    dz = pcoord->dx1f[i].item<double>();
    thermo_x->extrapolate_ad(temp, pres, xfrac, grav, dz);
  }

  // isothermal extrapolation
  for (; i <= ie; ++i) {
    auto mu = (thermo_x->mu * xfrac).sum(-1);
    dz = pcoord->dx1f[i].item<double>();
    pres *= exp(-grav * mu * dz / (kintera::constants::Rgas * temp));
    auto conc = thermo_x->compute("TPX->V", {temp, pres, xfrac});
    w[IPR].select(2, i) = pres;
    w[IDN].select(2, i) = thermo_x->compute("V->D", {conc});
    w.narrow(0, ICY, ny).select(3, i) = thermo_x->compute("X->Y", {xfrac});
  }

  // add noise
  w[IVX] += 0.01 * torch::rand_like(w[IVX]);
  w[IVY] += 0.01 * torch::rand_like(w[IVY]);

  // initialize
  torch::OrderedDict<std::string, torch::Tensor> vars;
  vars.insert("hydro_w", w);
  block->initialize(vars);

  // user output variables
  // (1) total precipitable mass fraction [kg/kg]
  block->user_out_var.insert("qtol", torch::Tensor());

  // output fields
  auto out2 = NetcdfOutput(
      OutputOptions().file_basename(exp_name).fid(2).variable("prim"));
  auto out3 = NetcdfOutput(
      OutputOptions().file_basename(exp_name).fid(3).variable("uov"));
  auto out4 = NetcdfOutput(
      OutputOptions().file_basename(exp_name).fid(4).variable("diag"));

  // create kinetics model
  auto op_kinet = kintera::KineticsOptions::from_yaml(infile);
  auto kinet = kintera::Kinetics(op_kinet);
  kinet->to(device);
  std::cout << fmt::format("Kinetics Options:\n{}", kinet->options)
            << std::endl;

  // time loop
  int count = 0;
  double current_time = 0.;
  while (!block->pintg->stop(count, current_time)) {
    auto dt = block->max_time_step(vars);

    // make output
    if (count % 20 == 0) {
      printf("count = %d, dt = %.6f, time = %.6f\n", count, dt, current_time);

      block->report_timer(std::cout);

      block->user_out_var["qtol"] = w.narrow(0, ICY, ny).sum(0);

      out2.write_output_file(block, vars, current_time, OctTreeOptions(), 0);
      out2.combine_blocks();
      out2.file_number++;

      out3.write_output_file(block, vars, current_time, OctTreeOptions(), 0);
      out3.combine_blocks();
      out3.file_number++;

      out4.write_output_file(block, vars, current_time, OctTreeOptions(), 0);
      out4.combine_blocks();
      out4.file_number++;
    }

    // evolve dynamics
    for (int stage = 0; stage < block->pintg->stages.size(); ++stage) {
      block->forward(dt, stage, vars);
    }

    // evolve kinetics
    auto &hydro_u = vars["hydro_u"];
    auto &hydro_w = vars["hydro_w"];

    auto temp = peos->compute("W->T", {hydro_w});
    auto pres = hydro_w[IPR];
    auto xfrac = thermo_y->compute("Y->X", {hydro_w.narrow(0, ICY, ny)});
    auto conc = thermo_x->compute("TPX->V", {temp, pres, xfrac});
    auto cp_vol = thermo_x->compute("TV->cp", {temp, conc});

    auto conc_kinet = kinet->options.narrow_copy(conc, thermo_y->options);
    auto [rate, rc_ddC, rc_ddT] = kinet->forward(temp, pres, conc_kinet);
    auto jac = kinet->jacobian(temp, conc_kinet, cp_vol, rate, rc_ddC, rc_ddT);
    auto del_conc = kintera::evolve_implicit(rate, kinet->stoich, jac, dt);
    std::vector<int64_t> vec(del_conc.dim(), 1);
    vec[del_conc.dim() - 1] = -1;
    auto del_rho = del_conc / thermo_y->inv_mu.narrow(0, 1, ny).view(vec);
    hydro_u.narrow(0, ICY, ny) += del_rho.permute({3, 0, 1, 2});

    count++;
    current_time += dt;
  }

  CommandLine::Destroy();
}

File: examples/example_crm.py
import torch
import math
import time
import numpy as np
import yaml
import torch
import snapy
import kintera
from snapy import (
        index,
        MeshBlockOptions,
        MeshBlock,
        OutputOptions,
        NetcdfOutput
        )
from kintera import (
        ThermoX,
        KineticsOptions,
        Kinetics,
        evolve_implicit
        )

torch.set_default_dtype(torch.float64)

def setup_moist_adiabatic_profile(config, coord, eos, thermo_x,
                                  device=torch.device("cpu")):
    Tmin = float(config["problem"]["Tmin"])
    grav = - float(config["forcing"]["const-gravity"]["grav1"])
    Ps = float(config["problem"]["Ps"])
    Ts = float(config["problem"]["Ts"])

    # dimensions
    nc3 = coord.buffer("x3v").shape[0]
    nc2 = coord.buffer("x2v").shape[0]
    nc1 = coord.buffer("x1v").shape[0]
    ny = len(thermo_x.options.species()) - 1
    nvar = eos.nvar()

    temp = Ts * torch.ones((nc3, nc2), device=device)
    pres = Ps * torch.ones((nc3, nc2), device=device)
    xfrac = torch.zeros((nc3, nc2, 1 + ny), device=device)

    # read in compositions
    for i in range(1, 1 + ny):
        name = 'x' + thermo_x.options.species()[i]
        if name in config["problem"]:
            xfrac[:, :, i] = float(config["problem"][name])

    # dry air mole fraction
    xfrac[:, :, 0] = 1.0 - xfrac[:, :, 1:].sum(dim=2)

    # adiabatic extrapolate half a grid
    ifirst = coord.ifirst()
    dx1f = coord.buffer("dx1f")
    dz = dx1f[ifirst].item()
    thermo_x.extrapolate_ad(temp, pres, xfrac, grav, dz / 2.0)

    nvapor = len(thermo_x.options.vapor_ids())
    ncloud = len(thermo_x.options.cloud_ids())
    i = ifirst
    w = torch.zeros((nvar, nc3, nc2, nc1), device=device)
    while i < coord.ilast():
        conc = thermo_x.compute("TPX->V", [temp, pres, xfrac])

        w[index.ipr, :, :, i] = pres
        w[index.idn, :, :, i] = thermo_x.compute("V->D", [conc])
        w[index.icy:, :, :, i] = thermo_x.compute("X->Y", [xfrac])

        if (temp < Tmin).any().item():
            raise ValueError("Temperature below minimum")
        dz = dx1f[i].item()
        thermo_x.extrapolate_ad(temp, pres, xfrac, grav, dz)
        i += 1

    # isothermal extrapolation
    while i < coord.ilast():
        mu = (thermo_x.buffer("mu") * xfrac).sum(-1);
        dz = dx1f[i].item()
        pres *= torch.exp(-grav * mu * dz / (kintera.constants.Rgas * temp))
        conc = thermo_x.compute("TPX->V", [temp, pres, xfrac])
        w[index.ipr, :, :, i] = pres
        w[index.idn, :, :, i] = thermo_x.compute("V->D", [conc])
        w[index.icy:, :, :, i] = thermo_x.compute("X->Y", [xfrac])
        i += 1

    # add noise
    w[index.ivx] += 0.01 * torch.rand_like(w[index.ivx])
    w[index.ivy] += 0.01 * torch.rand_like(w[index.ivy])

    return w;

def evolve_kinetics(block_vars, eos, thermo_x, thermo_y, kinet, dt):
    # evolve kinetics
    hydro_u = block_vars["hydro_u"]
    hydro_w = block_vars["hydro_w"]

    temp = eos.compute("W->T", [hydro_w])
    pres = hydro_w[index.ipr]
    xfrac = thermo_y.compute("Y->X", [hydro_w[index.icy:, :, :, :]])
    conc = thermo_x.compute("TPX->V", [temp, pres, xfrac])
    cp_vol = thermo_x.compute("TV->cp", [temp, conc])

    conc_kinet = conc[:, :, :, 1:]
    rate, rc_ddC, rc_ddT = kinet.forward_nogil(temp, pres, conc_kinet)
    jac = kinet.jacobian(temp, conc_kinet, cp_vol, rate, rc_ddC, rc_ddT)
    del_conc = evolve_implicit(rate, kinet.buffer("stoich"), jac, dt)
    del_rho = del_conc / thermo_y.buffer("inv_mu")[1:].view(1, 1, 1, -1)
    hydro_u[index.icy:, :, :, :] += del_rho.permute(3, 0, 1, 2)

if __name__ == "__main__":
    infile = "jupiter3d.yaml"
    config = yaml.safe_load(open(infile, "r"))

    # device
    device = torch.device("cuda:0")
    #device = torch.device("cpu")

    # set hydrodynamic options
    op = MeshBlockOptions.from_yaml(infile)
    block = MeshBlock(op)
    block.to(device)

    # get handles to modules
    coord = block.hydro.module("coord")
    thermo_y = block.hydro.module("eos.thermo")
    eos = block.hydro.get_eos()
    #thermo_y.options.max_iter(100)

    thermo_x = ThermoX(thermo_y.options)
    thermo_x.to(device)

    block_vars = {}
    interior = block.part((0, 0, 0))

    if "init_cond" in config["problem"]:
        nc3 = coord.buffer("x3v").shape[0]
        nc2 = coord.buffer("x2v").shape[0]
        nc1 = coord.buffer("x1v").shape[0]
        nvar = eos.nvar()
        block_vars["hydro_w"] = torch.zeros((nvar, nc3, nc2, nc1),
                                            device=device)

        module = torch.jit.load(config["problem"]["init_cond"])
        data = {name: param for name, param in module.named_buffers()}
        block_vars["hydro_w"][interior] = data["hydro_w"].to(device)
    else:
        block_vars["hydro_w"] = setup_moist_adiabatic_profile(
                config, coord, eos, thermo_x, device=device)

    # initialize
    block_vars = block.initialize(block_vars)

    out2 = NetcdfOutput(OutputOptions().file_basename("jupiter3d").fid(2).variable("prim"))
    out3 = NetcdfOutput(OutputOptions().file_basename("jupiter3d").fid(3).variable("uov"))
    out4 = NetcdfOutput(OutputOptions().file_basename("jupiter3d").fid(4).variable("diag"))

    # kinetics model
    op_kinet = KineticsOptions.from_yaml(infile)
    kinet = Kinetics(op_kinet)
    kinet.to(device)

    # integration
    count, current_time, start_time = 0, 0, time.time()
    while not block.intg.stop(count, current_time):
        dt = block.max_time_step(block_vars)

        # make output
        if count % 1000 == 0:
            print(f"count = {count}, dt = {dt}, time = {current_time}", flush=True)
            u = block_vars["hydro_u"]
            print("mass = ", u[interior][index.idn].sum(), flush=True)

            qtol = block_vars["hydro_w"][index.icy:, :, :, :].sum(dim=0)
            block.set_uov("qtol", qtol)

            for out in [out2, out3, out4]:
                out.increment_file_number()
                out.write_output_file(block, block_vars, current_time)
                out.combine_blocks()

        # evolve dynamics
        for stage in range(len(block.intg.stages)):
            block.forward(dt, stage, block_vars)

        evolve_kinetics(block_vars, eos, thermo_x, thermo_y, kinet, dt)

        count += 1
        current_time += dt
